### ISSUE TITLE
Tracking issue for #[link(kind)] connecting libraries on Windows

### ISSUE THREAD

**Issue by alexcrichton** [rust-lang/rust#37403]
Tracking issue for https://github.com/rust-lang/rfcs/pull/1717

- [x] Implement `-l[kind=][oldname:]libname`.
- [x] Implement lib<->symbol association and emit `dllimport` for dylib symbols.
- [x] Implement `kind="static-nobundle"`.

**Comment by dgrunwald** [comment]
If I specify `#[link(name="pythonXY")]` on 30 `extern C` blocks, and then compile with `rustc -l dylib=pythonXY:python27`, then `python27.lib` is passed to the linker 30 times.

Not sure if that's actually a problem, but maybe Rust could de-duplicate the linker arguments?

**Comment by dgrunwald** [comment]
Also, is it intentional that these changes are insta-stable?

cc @vadimcn

**Comment by vadimcn** [comment]
@dgrunwald: Could you coalesce those 30 extern blocks into one?  You probably did it to keep extern declarations in the same file/module as the code using them, but this isn't really how we thought about `extern` blocks when designing Rust FFI.

The problem with de-duping is that most unix linkers resolve references using only libraries to the right on the command line.  So if you know that `lib1` depends on something in `lib2`, they must appear in that order (even if `lib2` has already appeared to the left of `lib1`).  

Our approach so far has been to simply emit libraries in the order they appear in the source, followed by the ones specified on the command line.   

In order to optimize this, we'd need to provide some way of specifying library dependencies (e.g. "libraries {attached to the same extern block | within the same module | within the same source file} are guaranteed to be given to the linker in the same order").  For starters, some motivated person would need to write an RFC :)

> Also, is it intentional that these changes are insta-stable?

Sort of.  The changes in handling of `dllimport` did not create any new syntax that could be feature gated.   We could still use a `#[feature(...)]` to toggle the new behavior, but it's a bit too subtle for my liking: There would be no error message telling the user that this is feature-gated; they'd simply get a linker error.

**Comment by dgrunwald** [comment]
`python3-sys` is structured in the same way as the python headers: one `.rs` file per `.h` file; with the declarations within the file in the same order as in the `.h` file.
Because the Python headers have lots of macros (which I manually translate into constants or `#[inline] fn`s), I frequently close the `extern "C"` block, define the macro functions/constants, and then re-open the `extern "C"` block. This approach means that `python3-sys` has 95 `#[link(name="pythonXY")] extern "C"` blocks in total.

When a new Python version is released, I diff the new Python headers against the old headers and apply the same changes to the Rust code (using `#[cfg]` attributes for new functions). This would be much more difficult if I had to keep everything in a single extern block.

**Comment by retep998** [comment]
@dgrunwald A similar situation will occur for me in `winapi = "0.3"` so I'm really hoping there can be some sort of solution to library duplication.

**Comment by dgrunwald** [comment]
The current logic seems to be: if both `#[link(X)]` and `-l X:Y` is specified, ignore the position of the `-l X:Y` command-line argument and use all the `#[link]` positions instead.

What if we just did this the other way around: if both are specified, ignore the `#[link]` positions and use the single `-l X:Y` position instead.

**Comment by vadimcn** [comment]
In any case, this is not a new issue and has nothing to do with dllimport annotations.  Let's continue this discussion in #38460.

**Comment by retep998** [comment]
Would it be possible to get some focus on bikeshedding the name of `kind=static-nobundle` and perhaps stabilizing it?

**Comment by cramertj** [comment]
Are there any remaining issues blocking stabilization here? @retep998 I see you have thoughts about the name of `kind=static-nobundle`. Does anyone else have opinions on this?

**Comment by nagisa** [comment]
`cdylib` is the `dylib` equivalent of `dylib-nobundle`, isn’t it? Perhaps we could name `static-nobundle` a `cstaticlib`?

**Comment by retep998** [comment]
The only idea I have is to rename `static` to `static-bundle` and deprecate `static` with a warning that you may actually want `static-nobundle`.

**Comment by vadimcn** [comment]
Maybe `staticref`, as in "a reference to a static library"?  
I could also see `cstaticlib` standing for "using C's static linkage model".

IMO, the current use `static`ain't too bad: after all, in C world, there's no analogues of "linking static library to a static library", so we may as well define it the way we like.

**Comment by vadimcn** [comment]
... another option might be `latestatIc`, for "late-bound static library".

**Comment by retep998** [comment]
>IMO, the current use staticain't too bad: after all, in C world, there's no analogues of "linking static library to a static library", so we may as well define it the way we like.

In the (Microsoft) C world there absolutely is. `#pragma comment(lib, "foo.lib")` in a library will ensure that `foo.lib` is linked whenever that library is linked, and works identically to our current kinds `dylib` and `static-nobundle`. `static` is the odd one out with weird bundling behavior which keeps tripping up people who wonder why `static` is unable to find the static libraries installed on their system.

**Comment by vadimcn** [comment]
I am not sure that MSVC's `#pragma comment lib` is a pinnacle of intuitive design either.  For all other linking artifacts, "static" means "include it in the output", so bundling into .rlibs does make some amount sense as well.   Not to mention that the world does not end with MSVC...

In any case, it's too late now to change what `static` means in Rust.   What's the next best one?

**Comment by Neurrone** [comment]
Hoping this gets fixed before the 2018 edition, as it is extremely confusing behaviour. Just wasted over an hour trying to understand why it wouldn't find static .libs when I've specified all paths correctly.

**Comment by retep998** [comment]
Someone on IRC was getting hit by https://github.com/rust-lang/rust/issues/26003 and `static-nobundle` would have avoided that issue and solved their problem. Because it isn't stable they instead had to do workarounds by linking to it with `kind=dylib` and putting the symbols in a separate extern block so they wouldn't get dllimport'd.

Does anyone have any examples of actual legitimate use cases where `static` is required and `static-nobundle` won't work?

**Comment by vadimcn** [comment]
@retep998: Maybe you could submit a stabilization RFC for "static-nobundle"? (probably renamed to something prettier)   
I've been meaning to do that, but $work keeps me busy and I keep forgetting.  :(

It's probably too late for changing the way `static` works, and anyway, the current implementation doesn't seem totally unreasonable.

**Comment by indygreg** [comment]
While I've never contributed to Rust before, I got bit by the lack of a stabilized `static-nobundle` (see linked issue before this comment) and was wondering what I can do to help get it stabilized. If someone points me in the direction of a similar stabilization effort or gives me a list of TODOs, I could potentially work on a few things.

FWIW I think differentiating `static` and `static-nobundle` is useful because of developer ergonomics. With `static` you will get a build failure if a referenced library is not found at crate build time. Failing fast is generally a good idea. Especially since Cargo currently doesn't do a great job of reporting where library dependencies are coming from. I think the current `static` behavior is a reasonable default. And I *think* it matches the behavior of `dylib`.

I view `static-nobundle` as a special case because in most cases, symbols/libraries should be available at crate build time. I don't think many people will be building crates that want to reference libraries that don't exist yet. Or if things are this way, it is a special *flavor* of crate that deserves to be distinguished from traditional crates in some way. And that way could be `static-nobundle`.

Another thought that came to mind is potentially introducing a separate *axis* for controlling library searching. Today, we conflate static/dynamic linking/symbol resolution behavior with library searching. Perhaps we should differentiate between the two? e.g. we provide a mechanism to explicitly control the linkage type (static/dynamic) and rules for finding that library (at crate/rlib build time or deferred). I don't have a strong grasp on how all this is implemented in Rust, so my idea could be completely off base. But distinguishing between the two areas is a problem that comes up in build systems and it might be useful for Rust to give people precise control over behavior.

Anyway, please let me know how I can help and I'll see what I can do!

**Comment by retep998** [comment]
>And I think it matches the behavior of dylib.

It does not. `dylib` is deferred to link time just the same as `static-nobundle`. `static` is the odd one out with this behavior.

If you want to link to a library that is on the system, you *have* to defer to link time because only the *linker* knows where libraries are installed on the system. Having `rustc` be responsible for finding the library only works when the crate itself is providing the library, generally by compiling it from source.

**Comment by jethrogb** [comment]
There's an ICE when using this feature from a build script (without enabling the feature): https://github.com/rust-lang/rust/issues/53881

**Comment by jethrogb** [comment]
Also I'd like to chime in and say that this is not only useful on Windows.

**Comment by manhnt9** [comment]
I'm having same issue when cross-compiling on Linux for Windows:

```sh
.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-pc-windows-gnu/lib/libcompiler_builtins-477d6ada7584f678.rlib" "-Wl,-Bdynamic" "-ladvapi32" "-lpythonXY" "-lpythonXY" "-lpythonXY" "-lpythonXY" "-lpythonXY" "-lpythonXY" "-lpythonXY" "-lpythonXY" "-lpythonXY" "-lpythonXY" "-lpythonXY" "-lpythonXY" "-lpythonXY" "-lpythonXY" "-lpythonXY" "-lpythonXY" "-lpythonXY" "-lpythonXY" "-lpythonXY" "-lpythonXY" "-lpythonXY" "-lpythonXY" "-lpythonXY" "-lpythonXY" "-lpythonXY" "-lpythonXY" "-lpythonXY" "-lpythonXY" "-lpythonXY" "-lpythonXY" "-lpythonXY" "-lpythonXY" "-lpythonXY" "-lpythonXY" "-lpythonXY" "-lpythonXY" "-lpythonXY" "-lpythonXY" "-lpythonXY" "-lpythonXY" "-lpythonXY" "-lpythonXY" "-lpythonXY" "-lpythonXY" "-lpythonXY" "-lpythonXY" "-lpythonXY" "-lpythonXY" "-lpythonXY" "-lpythonXY" "-lpythonXY" "-lpythonXY" "-lpythonXY" "-lpythonXY" "-lpythonXY" "-lpythonXY" "-lpythonXY" "-lpythonXY" "-lpythonXY" "-lpythonXY" "-lpythonXY" "-lpythonXY" "-lpythonXY" "-lpythonXY" "-lpythonXY" "-lpythonXY" "-lpythonXY" "-lpythonXY" "-lpythonXY" "-lpythonXY" "-lpythonXY" "-lpythonXY" "-lpythonXY" "-lpythonXY" "-lpythonXY" "-lpythonXY" "-lpythonXY" "-lpythonXY" "-lpythonXY" "-lpythonXY" "-lpythonXY" "-lpythonXY" "-lpythonXY" "-lpythonXY" "-lpythonXY" "-lpythonXY" "-lpythonXY" "-lpythonXY" "-lpythonXY" "-lpythonXY" "-lpythonXY" "-lpythonXY" "-lpythonXY" "-lpython3.7m" "-lwinapi_advapi32" "-lwinapi_cfgmgr32" "-lwinapi_credui" "-lwinapi_kernel32" "-lwinapi_secur32" "-lwinapi_user32" "-ladvapi32" "-lws2_32" "-luserenv" "-Wl,-Bstatic" "-lgcc_eh" "-lpthread" "-Wl,-Bdynamic" "-lmingwex" "-lmingw32" "-lgcc" "-lmsvcrt" "-lmsvcrt" "-luser32" "-lkernel32" "/home/manh/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-pc-windows-gnu/lib/rsend.o"
  = note: /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          /usr/bin/x86_64-w64-mingw32-ld: cannot find -lpythonXY
          collect2: error: ld returned 1 exit status
```

**Comment by bbqsrc** [comment]
Is there a reason this hasn't been stabilised other than the name?

**Comment by petrochenkov** [comment]
I have submitted an RFC about re-branding "nobundle" as a modifier for the `static` linking kind - https://github.com/rust-lang/rfcs/pull/2951.

**Comment by petrochenkov** [comment]
I also have a question - does anyone actually use the library renaming (the `:rename` bit in `-lkind=name:rename`) in practice?

**Comment by indygreg** [comment]
> I also have a question - does anyone actually use the library renaming (the `:rename` bit in `-lkind=name:rename`) in practice?

I think I do!

The `pyembed` crate depends on a `libpython`, whose bindings are provided by a separate crate (currently `python3-sys` soon to be `pyo3`). The crate providing the bindings is the one emitting the `cargo:rustc-link-*` lines to link `libpython`.

Here's the wrinkle and my use case. The name of the actual Python library being linked is dynamic. And, the `pyembed` crate's build script needs to access metadata about the Python interpreter/library that `python3-sys` / `pyo3`'s build script resolves.

The `pyembed` crate's `Cargo.toml` defines a `links = "..."` to enable its build script to access `DEP_*` variables exported by `python3-sys` / `pyo3`'s build script so metadata can be exchanged between the build scripts (https://doc.rust-lang.org/cargo/reference/build-scripts.html#the-links-manifest-key).

However, because the actual library name is dynamic and sniffed out by the `python3-sys` / `pyo3` build scripts, the `pyembed` crate can't use a static value for the `links = "..."` config to represent the _real_ library! So, the workaround is for `python3-sys` / `pyo3` to define their `extern "C"` bindings as linking against `pythonXY` and have their build scripts emit the `cargo:rustc-link-lib=pythonXY:real_library` line to alias `pythonXY` to the dynamically discovered library name. `pyembed` uses `links = "pythonXY"`, enabling the `DEP_*` variables to be defined and for metadata to be exchanged.

There is a wonky edge case where when using `static-nobundle` something somewhere insists on a `pythonXY` static library actually existing. I'm unsure if this is wonkiness in Cargo or how these particular crates are configured. But I work around it by writing out a placeholder `pythonXY` static library somewhere on the link path and the error goes away: the linker only cares that it can find the symbols, not necessarily that the symbols are defined in the exact library name annotated in an `extern "C"` block.

I recognize this use case is kinda wonky. If there were a better way to pass metadata from one build script to a dependent build script, I'd do that. But this `links = "..."` mechanism is what I've settled on.

**Comment by indygreg** [comment]
... and a few hours later and I've refactored the `pyembed` crate to no longer depend on `static-nobundle`. I did this by teaching `pyo3`'s build script to read a config file (specified via an environment variable that must be defined outside of `cargo`) which defines (among other things) a list of extra lines to `println!()` from the build script. The `pyo3` build script now emits a _complete_ definition of the static library it needs to link against, removing the need for the deferred symbol resolution via `static-nobundle`. Hacky. But it gets the job done.

I was fortunate to be working with a crate maintainer receptive to hacking up their build script to support my use case. Not everyone will be so fortunate, so this feature is still generally useful.

**Comment by pnkfelix** [comment]
From quickly skimming the comments here, I can tell there are people invested in this feature, but I cannot tell what its current status is (how much has been stabilized, how much was insta-stabilized, how much is still unstable), nor how much remains to be done, if anything.

@rustbot label S-tracking-needs-summary

**Comment by petrochenkov** [comment]
This issue is basically tracking the removal of `static-nobundle` now.

It's currently blocked on https://github.com/rust-lang/rust/pull/93901 and on one more following PR stabilizing the `bundle` modifier.

**Comment by CoorFun** [comment]
Hi there:

I am pointed here by [this post](https://users.rust-lang.org/t/linking-against-c-dependencies/25099/4), what I am actually trying to do now is that I want to find a way to link my rust final c dynamic library (.so) against stdlibc++ statically. So that this shared library can be used under another linux distribution which doesn't have the minimum requried stdlibc++ installed.. 

From the post, I fount that "static-nobundle" feature could help, but when I tried it in my lib, it shows me: 
```text
warning: library kind `static-nobundle` has been superseded by specifying modifier `-bundle` with library kind `static`. Try `static:-bundle`
```

So I changed `println!("cargo:rustc-link-lib=static-nobundle=stdc++")`to  `println!("cargo:rustc-link-lib=static:-bundle=stdc++")` in the `build.rs` file

But still, I am not getting things compiled as expected:
```text
error: linking modifier `bundle` is unstable and only accepted on the nightly compiler
```

I have enabled those two features in my `lib.rs`, and the toolchain used is `nightly-2022-05-24-x86_64-unknown-linux-gnu`

lib.rs:
```rust
#![feature(native_link_modifiers_bundle)]
#![feature(static_nobundle)]
```

Anyone has ideas? Thanks in advance!

**Comment by AlexTMjugador** [comment]
> Anyone has ideas? Thanks in advance!

Using `static:-bundle=stdc++` worked for me on `rustc 1.79.0-nightly (0bf471f33 2024-04-13)` without enabling any feature flags. The rollup merge that closed this issue merged this PR, which stabilized the new linker modifier syntax: https://github.com/rust-lang/rust/pull/95818

**Comment by gucki** [comment]
@AlexTMjugador I can't get my cross-compiled executable to work. Running it always fails with "The code execution cannot proceed because libstdc++-6.dll was not found...". 

This is my build.rs:

    fn main() {
      println!("cargo:rustc-link-lib=static:-bundle=stdc++");
    }

This is my docker image I use for the build:

    FROM rustlang/rust:nightly

    RUN apt update
    RUN apt install -y mingw-w64 cmake
    RUN update-alternatives --set x86_64-w64-mingw32-g++ /usr/bin/x86_64-w64-mingw32-g++-posix
    RUN update-alternatives --set x86_64-w64-mingw32-gcc /usr/bin/x86_64-w64-mingw32-gcc-posix
    RUN update-alternatives --set i686-w64-mingw32-g++ /usr/bin/i686-w64-mingw32-g++-posix
    RUN update-alternatives --set i686-w64-mingw32-gcc /usr/bin/i686-w64-mingw32-gcc-posix

    RUN rustup target add x86_64-pc-windows-gnu

    WORKDIR /build

Output of "cargo version --verbose" inside the container:

    cargo 1.80.0-nightly (6087566b3 2024-04-30)
    release: 1.80.0-nightly
    commit-hash: 6087566b3fa73bfda29702632493e938b12d19e5
    commit-date: 2024-04-30
    host: x86_64-unknown-linux-gnu
    libgit2: 1.7.2 (sys:0.18.3 vendored)
    libcurl: 8.6.0-DEV (sys:0.4.72+curl-8.6.0 vendored ssl:OpenSSL/1.1.1w)
    ssl: OpenSSL 1.1.1w  11 Sep 2023
    os: Debian 12.0.0 [64-bit]

Do you know why the library is not statically linked?

**Comment by AlexTMjugador** [comment]
> Do you know why the library is not statically linked?

My best guess as to why that occurred is that another thing is trying to dynamically link to the C++ stdlib. Most likely, that thing is another build script, or a configured linker flag. You can try building your project with Cargo's verbose flags to pinpoint the source of the troublesome linker parameters.

I was able to verify that the `cargo:rustc-link-lib=static:-bundle=stdc++` instruction works as expected to link to the C++ stdlib. I spun up a fresh Docker container with your Dockerfile and newly created Cargo project with `cargo new linking-test`, appending the `build.rs` you provided to it. The output of `x86_64-w64-mingw32-objdump` confirms that the executable should not require `libstdc++-6.dll` at runtime:

```
root@b8083609ca3d:~/linking-test# x86_64-w64-mingw32-objdump -p target/x86_64-pc-windows-gnu/debug/linking-test.exe | grep 'DLL Name'
        DLL Name: KERNEL32.dll
        DLL Name: msvcrt.dll
        DLL Name: ntdll.dll
        DLL Name: USERENV.dll
        DLL Name: WS2_32.dll
        DLL Name: api-ms-win-core-synch-l1-2-0.dll
        DLL Name: bcryptprimitives.dll
```

**Comment by mati865** [comment]
That is right, libLLVM is linked statically to libstdc++.
